src/BuiltInTools/dotnet-watch/Internal/HotReloadFileSetWatcher.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Tools.Internal;
+
+namespace Microsoft.DotNet.Watcher.Internal
+{
+    public class HotReloadFileSetWatcher : IDisposable
+    {
+        private static readonly TimeSpan DebounceInterval = TimeSpan.FromMilliseconds(50);
+        private readonly FileWatcher _fileWatcher;
+        private readonly FileSet _fileSet;
+        private TaskCompletionSource<FileItem[]?>? _tcs;
+        private bool _initialized;
+        private bool _disposed;
+        private volatile ConcurrentDictionary<string, FileItem> _changedFiles = new(StringComparer.Ordinal);
+        private volatile ConcurrentDictionary<string, FileItem> _swap = new(StringComparer.Ordinal);
+
+        public HotReloadFileSetWatcher(FileSet fileSet, IReporter reporter)
+        {
+            Ensure.NotNull(fileSet, nameof(fileSet));
+
+            _fileSet = fileSet;
+            _fileWatcher = new FileWatcher(reporter);
+        }
+
+        private void EnsureInitialized()
+        {
+            if (_initialized)
+            {
+                return;
+            }
+
+            _initialized = true;
+
+            foreach (var file in _fileSet)
+            {
+                _fileWatcher.WatchDirectory(Path.GetDirectoryName(file.FilePath));
+            }
+
+            _fileWatcher.OnFileChange += FileChangedCallback;
+
+            Task.Factory.StartNew(async () =>
+            {
+                // Debounce / polling loop
+                while (!_disposed)
+                {
+                    await Task.Delay(DebounceInterval);
+                    if (_changedFiles.Count == 0)
+                    {
+                        continue;
+                    }
+
+                    var tcs = Interlocked.Exchange(ref _tcs, null!);
+                    if (tcs is null)
+                    {
+                        continue;
+                    }
+
+                    // Replace the changed file collection as we're taking a snapshot of the batch.
+                    _swap.Clear();
+                    var changedFiles = Interlocked.Exchange(ref _changedFiles, _swap);
```

Comment:
Can we just lock instead of interlocked? This isn't perf sensitive code.

Reply:
Done. Using a lock + a single dictionary. I imagine I could turn it in to a regular dictionary, now that it doesn't need to be concurrent, but it feels like a no biggie.